---
title: "Benchmarking scry methods with DelayedArray backends"
author: "Stephanie Hicks"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  BiocStyle::html_document:
    toc: false
<!-- vignette: > -->
<!--   %\VignetteEngine{knitr::knitr} -->
<!--   %\VignetteIndexEntry{Benchmarking scry methods with DelayedArray backends} -->
<!--   %\usepackage[UTF-8]{inputenc} -->
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Make sure you have latest version (needs glmpca >= 0.2.0)
```{r, eval=FALSE}
remotes::install_github("willtownes/glmpca")
remotes::install_github("kstreet13/scry@hdf5") # install HDF5 branch
```


```{r}
suppressPackageStartupMessages({
  library(here)
  library(SingleCellExperiment)
  library(DelayedArray)
  library(BiocSingular)
  library(BiocParallel)
  library(scry)
})
```

# Overview

For large data sets, it may be advantageous to keep the original data in a on-disk HDF5 file and avoid loading the full data in memory.

This is a vignette to benchmark how `scry` performs with in-memory data vs matrix-like objects from [`DelayedArray`](https://bioconductor.org/packages/DelayedArray) Bioconductor package (`DelayedMatrix`, `HDF5Matrix`). 

For now, we are focused on benchmarking the `scry::nullResiduals` function with a Poisson model and using deviance residuals. We will expand this to other options later. 

To get started, we will work with the `TENxBrainData` ExperimentHub data package that contains 1.3 million brain cell scRNA-seq data from 10X Genomics. 

```{r}
library(TENxBrainData)
```

# Simple benchmark

```{r}
tenx <- TENxBrainData()
tenx <- tenx[,1:1000] # start with a small subset for testing while building benchmark code
```

First let's realize the counts matrix in the `tenx` SCE. 
```{r}
tenx_realized <- tenx
assay(tenx_realized, "counts") <- as.matrix(assay(tenx, "counts"))
assay(tenx_realized, "counts")[1:5, 1:5]
dim(tenx_realized)
```

Next, confirm that the sizes are very different: 
```{r}
pryr::object_size(tenx) # 4.62 MB
pryr::object_size(tenx_realized) # 117 MB
```

Yep, they are different. OK let's apply the various options: 

Here, we run the in-memory, vectorized `nullResiduals()` function: 

```{r}
time_start <- proc.time()
tenx_realized <- nullResiduals(tenx_realized, assay="counts",
                               fam = "poisson", type="deviance")
pcs_out <- runPCA(assay(tenx_realized, "poisson_deviance_residuals"), 
              rank=10, BSPARAM=FastAutoParam(), get.rotation=FALSE)
time_end <- proc.time()
time_inmem_vec <- time_end - time_start
time_inmem_vec
```

Next, we try the `nullResiduals()` function using on-disk data (`DelayedArray`). 
The `nullResiduals()` function that uses `BiocSingular::LowRankMatrix`. 
```{r}
time_start <- proc.time()
tenx <- nullResiduals(tenx, assay="counts", 
                     fam = "poisson", type="deviance")
pcs_out <- runPCA(assay(tenx, "poisson_deviance_residuals"), 
                  rank=10, BSPARAM=FastAutoParam(), get.rotation=FALSE)
time_end <- proc.time()
time_da_lrm <- time_end - time_start
time_da_lrm

str(pcs_out)
```

Compare the times 
```{r}
time_inmem_vec
time_da_lrm
# > time_inmem_vec
#   user  system elapsed 
#  9.663   0.838  10.542 
# > time_da_lrm
#    user  system elapsed 
# 751.583 119.311 880.660 
```


```{r}
sessionInfo()
```

